component/core.py

```diff
@@ -400,23 +406,26 @@ def component(self, usage=None, model_name=None):
                 if c.apply_on_models and model_name in c.apply_on_models
             ]
         if len(component_classes) != 1:
-            raise SeveralComponentError(
-                "Several components found for collection '%s', "
-                "usage '%s', model_name '%s'. Found: %r"
-                % (
-                    self.collection._name,
-                    usage or "",
-                    model_name or "",
-                    component_classes,
+            if first:
+                component_classes = component_classes[:1]
```

Comment:
Still a bit puzzled by this after our discussion. This is not deterministic so is it really good?
Also, on the fact that you don't want to use `many_components()[0]`, did you actually measure the cost of the instantiation of the components? 
We shouldn't sacrifice API's understandability if the optimization is negligible.
In this case, I think it makes the interface less clear (different exceptions depending on parameter, not deterministic result).

An alternative would be to add a private method

```python
def _matching_components(self, usage=None, model_name=None, **kw):
        if isinstance(model_name, models.BaseModel):
            model_name = model_name._name
        model_name = model_name or self.model_name
        component_classes = self._lookup_components(
            usage=usage, model_name=model_name, **kw
        )
        if model_name == self.model_name:
            work_context = self
        else:
            work_context = self.work_on(model_name)
        return component_classes, work_context
```

This method can be used in `components` and `many_components`, but also from implementations which can filter again the component_classes. But again, I have the feeling the instantiation is negligible...

Reply:
yeah, I was typing something similar (`_query_components` in my case) but then I decided to keep it simple and not change too much what is there today.
Regarding the cost of init: well, you don't know. It really depends on what ppl do in their init if they customized it.
For example: I added a validation of the work ctx attributes whereas if a required attribute is not there an exception is raised. In this case, I don't won't that piece of code to be awaken if is not needed.

Reply:
> not change too much what is there today.

In my head, adding such parameter is a heavier change than extracting the method `_query_components` :shrug:.

What about the fact that the component returned is not deterministic? What's the idea?

Reply:
@guewen should be better now. `component` is left untouched (apart moving out some utils) and `query_component` does the job for the fault tolerant lookup.